rust/stackable-cockpit/src/oci.rs

```diff
@@ -0,0 +1,173 @@
+use std::collections::HashMap;
+
+use serde::Deserialize;
+use snafu::{OptionExt, ResultExt, Snafu};
+use tracing::debug;
+use urlencoding::encode;
+
+use crate::{
+    constants::{HELM_OCI_BASE, HELM_REPO_NAME_DEV, HELM_REPO_NAME_STABLE, HELM_REPO_NAME_TEST},
+    utils::chartsource::{ChartSourceEntry, ChartSourceMetadata},
+};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("cannot get repositories"))]
+    GetRepositories { source: reqwest::Error },
+
+    #[snafu(display("cannot parse repositories"))]
+    ParseRepositories { source: reqwest::Error },
+
+    #[snafu(display("cannot get artifacts"))]
+    GetArtifacts { source: reqwest::Error },
+
+    #[snafu(display("cannot parse artifacts"))]
+    ParseArtifacts { source: reqwest::Error },
+
+    #[snafu(display("unexpected OCI repository name"))]
+    UnexpectedOciRepositoryName,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct OciRepository {
+    pub name: String,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct Tag {
+    pub name: String,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct Artifact {
+    pub digest: String,
+    pub tags: Option<Vec<Tag>>,
+}
```

Comment:
note: Please add doc comments for these "model" structs and shorty mention what they are describing. Maybe also add a small excerpt of the JSON they are deserializing from.

Reply:
dce1320